### PR TITLE
Use semester globally

### PR DESCRIPTION
**Pull request by victor-paumier**
Remove link between semester & section to make semesters common to all sections

### CODE REVIEW

File: src/Command/InitPreviousSemestersCommand.php [hunk 1]
```diff
@@ -43,38 +46,34 @@ public function __construct(
...
-                        } else {
-                            $semester->setActivated(false);
-                        }
-                    }
-
-                    $this->em->persist($semester);
-                }
-                ++$year;
+            $semester = $section->getSemester();
+            $year = (1 == $semester)
```
Comment (JeremieSamson) [inline src/Command/InitPreviousSemestersCommand.php]: parentesis not needed

File: src/Command/InitPreviousSemestersCommand.php [hunk 2]
```diff
@@ -43,38 +46,34 @@ public function __construct(
...
+                ? '2019 - 2020'
+                : '2018 - 2019'
+            ;
+
+            $currentSemester = $this->semesterRepo->findOneBy(['year' => $year, 'period' => $this->periodChoices[$semester - 1]]);
```
Comment (JeremieSamson) [inline src/Command/InitPreviousSemestersCommand.php]: can you multiline here plz ?

File: src/DataFixtures/SemesterFixtures.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+use App\Entity\Semester;
+use App\Enum\SemesterPeriodEnum;
+use Doctrine\Bundle\FixturesBundle\Fixture;
+use Doctrine\Common\DataFixtures\OrderedFixtureInterface;
+use Doctrine\Common\Persistence\ObjectManager;
+
+class SemesterFixtures extends Fixture implements OrderedFixtureInterface
+{
+    const SEMESTER_REFERENCE = 'semester';
```
Comment (JeremieSamson) [inline src/DataFixtures/SemesterFixtures.php]: visibility plz

File: src/Repository/SemesterRepository.php [hunk 4]
```diff
@@ -19,16 +19,34 @@ public function getDistinctYears(Section $section = null): array
...
         if ($section) {
-            $qb->andWhere('s.section = :section')
-                ->setParameter('section', $section)
-            ;
+            $subQuery = $this->createQueryBuilder('semester')
+                ->select('semester.id')
+                ->innerJoin('semester.users', 'users')
+                ->innerJoin('users.section', 'section')
+                ->where('section.id = :section')
+                ->getQuery()->getDQL();
```
Comment (JeremieSamson) [inline src/Repository/SemesterRepository.php]:
```suggestion
                ->getQuery()->getDQL()
            ;
```

File: src/Repository/SemesterRepository.php [hunk 5]
```diff
@@ -19,16 +19,34 @@ public function getDistinctYears(Section $section = null): array
...
+
+            $qb
+                ->where($qb->expr()->in('s.id', $subQuery))
+                ->setParameter('section', $section->getId())
+                ;
```
Comment (JeremieSamson) [inline src/Repository/SemesterRepository.php]:
```suggestion
            ;
```

File: src/Repository/SemesterRepository.php [hunk 6]
```diff
@@ -19,16 +19,34 @@ public function getDistinctYears(Section $section = null): array
...
         }, $qb->getQuery()->getArrayResult());
     }
+
+    public function getNbAvailableChoices(Section $section): int
+    {
+        return $this->createQueryBuilder('s')
+            ->select('count(s.id)')
+            ->where('s.id > :idSemester')
+            ->setParameter('idSemester', $section->getCurrentSemester()->getId())
+            ->getQuery()->getSingleScalarResult();
```
Comment (JeremieSamson) [inline src/Repository/SemesterRepository.php]:
```suggestion
            ->getQuery()
            ->getSingleScalarResult()
        ;
```

File: src/Migrations/Version20190513125750.php [hunk 7]
```diff
@@ -36,20 +36,13 @@ public function up(Schema $schema): void
           second_semester_end second_semester_end DATE NOT NULL');
         $this->addSql('CREATE TABLE semester (
           id INT AUTO_INCREMENT NOT NULL, 
-          section_id INT DEFAULT NULL,
```
Comment (JeremieSamson) [inline src/Migrations/Version20190513125750.php]: IMO, we should follow good practices and not change previous migrations but create a new one with changes.

File: src/Services/MailManager.php [hunk 8]
```diff
@@ -124,10 +124,11 @@ public function notifyBCSwitchSemester(Section $section): int
     {
         $semester = $section->getCurrentSemester();
         $bcs = $this->userRepo->getGrantedBC($section);
+        $mailSent = 0;
```
Comment (JeremieSamson) [inline src/Services/MailManager.php]: IMO var should be `$mailSentCount` or something like that, WDYT ?

File: src/Services/SemesterHelper.php [hunk 9]
```diff
@@ -80,47 +81,55 @@ public function switchSemester(Section $section): void
         $this->statsService->generateNewStats($section, true);
     }
 
-    public function generateNewYearSemester(Section $section, int $baseYear): array
+    public function generateNewYearSemester(int $baseYear, bool $andFlush = false): array
```
Comment (JeremieSamson) [inline src/Services/SemesterHelper.php]: just `$flush` here maybe ?

Comment (victor-paumier) [reply src/Services/SemesterHelper.php]: I first used this variable with fosUserMessage and as they named it `$andFlush` I used the same name

**Pull request merged**